lib/ts/recipe/oauth2/types.ts

```diff
@@ -0,0 +1,179 @@
+    recipeImplementation: RecipeInterface;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type RecipeInterface = {
+    tokensPOST(input: {
+        payload?: any;
+        validitySeconds?: number;
+        useStaticSigningKey?: boolean;
+        userContext: any;
+    }): Promise<
+        | {
+              token_type: "Bearer";
+              access_token: string;
+              refresh_token?: string;
+              id_token?: string;
+              expires_in: number;
+          }
+        | GeneralErrorResponse
+    >;
+    // The OpenId spec states that we have to support both POST and GET
+    authorizeGET(input: {
+        clientId: string;
+        responseType: string;
+        redirectUri?: string;
+        scopes: string[];
+        state?: string;
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: string;
+
+        // OpenIdConnect
+        prompt?: string;
+        nonce?: string;
+
+        // Not sure if we want to support these
+        acr_values?: string;
+        maxAge?: number;
+
+        options: APIOptions;
+        userContext: any;
+    }): Promise<{
+        redirectUrl: string; // This will end up as a 302, encoding both error and success
+    }>;
+    authorizePOST(input: {
+        clientId: string;
+        responseType: string;
+        redirectUri?: string;
+        scopes: string[];
+        state?: string;
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: string;
+
+        // OpenIdConnect
+        prompt?: string;
+        nonce?: string;
+
+        // Not sure if we want to support these
+        acr_values?: string;
+        maxAge?: number;
+
+        options: APIOptions;
+        userContext: any;
+    }): Promise<{
+        redirectUrl: string; // This will end up as a 302, encoding both error and success
+    }>;
+    // This is intended to be used when we want to do the redirection in the FE (e.g.: header based auth)
+    getRedirectUrlPOST(input: {
+        clientId: string;
+        responseType: string;
+        redirectUri?: string;
+        scopes: string[];
+        state?: string;
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: string;
+
+        // OpenIdConnect
+        prompt?: string;
+        nonce?: string;
+
+        // Not sure if we want to support these
+        acr_values?: string;
+        maxAge?: number;
+
+        options: APIOptions;
+        userContext: any;
+    }): Promise<
+        | {
+              status: "OK";
+              redirectUrl: string;
+          }
+        | GeneralErrorResponse
+    >;
+
+    userInfoGET(input: {
+        options: APIOptions;
+        userContext: any;
+    }): Promise<{ status: "OK"; sub: string; email?: string; email_verified?: boolean; phoneNumber?: string }>; // We should define this in a way that doesn't require adding all claims here
```

Comment:
> We should define this in a way that doesn't require adding all claims here

I didn't understand

Reply:
I meant that we should define the result of the function in a way that:
- will not require us to change it in case we add a built-in scope that adds some info int the user info
- will make it easy to override (TS could be complaining about types if we make it too strict)
For now, I made it into `JSONObject`; it should fit any object that we can be serialize without an issue.